pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java

```diff
@@ -163,11 +171,10 @@ void runCmd() throws Exception {
         protected String deserializationClassName;
         @Parameter(names = "--parallelism", description = "The source's parallelism factor (i.e. the number of source instances to run)")
         protected Integer parallelism;
-        @Parameter(
-                names = "--jar",
-                description = "Path to the jar file for the Source",
-                listConverter = StringConverter.class)
+        @Parameter(names = "--jar", description = "Path to the jar file for the Source. It also supports url-path (http/https/file) from which worker can download the package.", listConverter = StringConverter.class)
         protected String jarFile;
+        @Parameter(names = "--classNameArgType", description = "Sink impl class's argument type if jar-file path is url")
```

Comment:
> We can already resolve the types of sink/source/functions from jars

Correct. However, right now function can also support `http and file` url. we can download jar from http and load the class but with `file` url, package is present at worker host file-system which can't be accessed by the client where cli is running. so, cli requires user to provide arg/return class-name that will be used [PulsarSink](https://github.com/apache/incubator-pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java?utf8=%E2%9C%93#L243) and source. I have added description that it only requires for file-url.

Reply:
If the file is present in the workers, then the client cannot do any kind of type validation. However the function worker should be able to and should reject anything out of order. Thus I still don't see the need for this

Reply:
> Thus I still don't see the need for this

Actually, In current function-admin-api cli-client passes:
[sinkSpecBuilder.setTypeClassName(typeArg)](https://github.com/rdhabalia/pulsar/blob/663ac3888cc3ce6b1f8c8d2423648575321ed8fb/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSinks.java#L397) and [sourceSpecBuilder.setTypeClassName(typeArg)](https://github.com/rdhabalia/pulsar/blob/663ac3888cc3ce6b1f8c8d2423648575321ed8fb/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java#L358) and server doesn't perform any validation.

if jar is present locally then cli loads and finds out sink/source arg-type but it's not the same case for file-url therefore, I have added argument  to [set arg-type](https://github.com/rdhabalia/pulsar/blob/663ac3888cc3ce6b1f8c8d2423648575321ed8fb/pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdSources.java#L330) if it's provided.

> However the function worker should be able to and should reject anything out of order. 

Correct, we should also perform server side validation which we are not doing right now and server stores arg-type what it receives in http-request. I will add it.

Reply:
I would rather add that instead of this type args which can be filled in by the server at submission time. My view is to limit the growth of functionconfig as much as possible.